hw/xbox/nv2a/nv2a_vsh.c

```diff
@@ -674,13 +674,25 @@ static const char* vsh_header =
     "#define EXP(dest, mask, src) dest.mask = _EXP(_in(src).x).mask\n"
     "vec4 _EXP(float src)\n"
     "{\n"
-    "  return vec4(exp2(src));\n"
+    "  vec4 result;\n"
+    "  result.x = exp2(floor(src));\n"
+    "  result.y = src - floor(src);\n"
+    "  result.z = exp2(src);\n"
+    "  result.w = 1.0;\n"
+    "  return result;\n"
     "}\n"
     "\n"
     "#define LOG(dest, mask, src) dest.mask = _LOG(_in(src).x).mask\n"
     "vec4 _LOG(float src)\n"
     "{\n"
-    "  return vec4(log2(src));\n"
+    "  float tmp = abs(src);\n"
+    "  if (tmp == 0.0) { return vec4(-INFINITY, 1.0f, -INFINITY, 1.0f); }\n"
+    "  vec4 result;\n"
+    "  result.x = floor(log2(tmp));\n"
+    "  result.y = tmp / exp2(floor(log2(tmp)));\n"
```

Comment:
When adding the bias in EXP, we might have to undo it here.
Would be nice if someone thought about the consequences and reported back what could go wrong here.